mne/io/reference.py

```diff
@@ -467,39 +476,53 @@ def set_bipolar_reference(inst, anode, cathode, ch_name=None, ch_info=None,
         raise ValueError('Number of channel info dictionaries must equal the '
                          'number of anodes/cathodes.')
 
-    # Merge specified and anode channel information dictionaries
-    new_chs = []
-    for ci, (an, ch) in enumerate(zip(anode, ch_info)):
-        _check_ch_keys(ch, ci, name='ch_info', check_min=False)
-        an_idx = inst.ch_names.index(an)
-        this_chs = deepcopy(inst.info['chs'][an_idx])
-
-        # Set channel location and coil type
-        this_chs['loc'] = np.zeros(12)
-        this_chs['coil_type'] = FIFF.FIFFV_COIL_EEG_BIPOLAR
-
-        this_chs.update(ch)
-        new_chs.append(this_chs)
-
     if copy:
         inst = inst.copy()
 
-    for i, (an, ca, name, chs) in enumerate(
-            zip(anode, cathode, ch_name, new_chs)):
-        if an in anode[i + 1:] or an in cathode[i + 1:] or not drop_refs:
-            # Make a copy of the channel if it's still needed later
-            # otherwise it's modified inplace
-            _copy_channel(inst, an, 'TMP')
-            an = 'TMP'
-        _apply_reference(inst, [ca], [an])  # ensures preloaded
-        an_idx = inst.ch_names.index(an)
-        inst.info['chs'][an_idx] = chs
-        inst.info['chs'][an_idx]['ch_name'] = name
-        logger.info('Bipolar channel added as "%s".' % name)
-        inst.info._update_redundant()
+    anode = _check_before_reference(inst, ref_from=cathode,
+                                    ref_to=anode, ch_type='auto')
+
+    # Create bipolar reference channels by multiplying the data (channels x
+    # time) with a matrix n_virtual_channels x channels
+    # and add them to the instance
+    multiplier = np.zeros((len(anode), len(inst.ch_names)))
+    for idx, (a, c) in enumerate(zip(anode, cathode)):
+        multiplier[idx, inst.ch_names.index(a)] = 1
+        multiplier[idx, inst.ch_names.index(c)] = -1
+
+    # Create new instances (create each from single channel to allow same
+    # channel in anode multiple times)
+    ref_instances = list()
+    for ch_idx, (an, ca, name, info) in enumerate(zip(anode, cathode,
+                                                      ch_name, ch_info)):
```

Comment:
couldn't you simply do:

```py
ref_inst = inst.copy().pick_channels(cathode)
ref_inst.rename_channels(...)
```

it will read cleaner and maybe even faster

Reply:
I did this at first, but this fails when one channel occurs multiple times in anode or cathode, as pick_channels converts to a set. It would be cleaner and probably faster, but at the cost, that you can only use each channel once on each orientation

Reply:
There is a test for this in test_reference which fails then

Reply:
ughgh ... I see. I would try to avoid using `add_channels` because that will probably make unnecessary copies of the data. Can you use `RawArray` or `EvokedArray` or some such thing to construct the final instance in one go? I suspect it will be faster

Reply:
>  I would try to avoid using add_channels because that will probably make unnecessary copies of the data (for each new channel?). Can you use RawArray or EvokedArray or some such thing to construct the final instance in one go? I suspect it will be faster

Indeed iterating will be slower than a vectorized solution. I would figure out what dot product needs to be done, then do a single `raw.drop_channels` call, then do a single `raw.add_channels` call with the new data. The overhead from this should be much smaller